modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java

```diff
@@ -174,6 +206,8 @@ private void writeServiceInvokeRequest(
                         writer.writeObject(args[i]);
                     }
                 }
+
+                writer.writeMap(withCallCtx ? ((ServiceCallContextImpl)callCtx).values() : null);
```

Comment:
What if the user implements and passes to method their own `ServiceCallContext` implementation? We should check it and throw an error with reasonable message, or introduce some kind of `attributeNames()` method in public interface, to get all possible attributes for this context.

Reply:
Also, should not write anything (including `null`) if `ProtocolBitmaskFeature.SERVICE_INVOKE_CALLCTX` is not supported.

Reply:
> What if the user implements and passes to method their own `ServiceCallContext` implementation? We should check it and throw an error with reasonable message, or introduce some kind of `attributeNames()` method in public interface, to get all possible attributes for this context.

Added explicit user arguments check.

> Also, should not write anything (including `null`) if `ProtocolBitmaskFeature.SERVICE_INVOKE_CALLCTX` is not supported.

Done.